cmd/access/node_builder/staked_access_node_builder.go

```diff
@@ -102,7 +116,7 @@ func (anb *StakedAccessNodeBuilder) Build() AccessNodeBuilder {
 
 				// register the proxy engine with the unstaked network
 				var err error
-				unstakedNetworkConduit, err = node.Network.Register(engine.PublicSyncCommittee, proxyEngine)
+				unstakedNetworkConduit, err = anb.AccessNodeConfig.PublicNetworkConfig.Network.Register(engine.PublicSyncCommittee, proxyEngine)
```

Comment:
🟥  **Red Flag**: Current implementation of `splitter.Engine` and `RequestHandlerEngine` is breaking a cardinal principle of Flow's software architecture: 

A [`Conduit`](https://github.com/onflow/flow-go/blob/master/network/conduit.go#L59-L64) is the _exclusive_ interface of an engine for interacting with the network, and each engine _must be_ provided a mutually exclusive conduit for networking interactions. From the perspective of the networking layer, each conduit _must be_ in the sole possession of one and _only one_ engine. Unique conduit(s) per engine is similar to the unique port(s) per each process in operating systems. 

However, the current implementation breaks this software architecture principle of Flow by sharing a conduit between two engines, where `splitter.Engine` reads from it, and `synchronization.RequestHandlerEngine` writes on it. 

Please refactor the architecture to make sure that a conduit is in the sole possession of only a single engine. You may refactor the `synchronization.RequestHandlerEngine` to make the conduit mutually exclusive to it. Then for ingress events, the `synchronization.RequestHandlerEngine` utilizes the `ProcessLocal` method of the  `splitter.Engine` to pass the events to it. In this way, the implementation observes the software architecture principles of Flow, and engines and conduits are utilized consistently with what they are meant for. 

A similar architecture has been implemented with [`requester`](https://github.com/onflow/flow-go/blob/master/engine/verification/requester/requester.go) and [`fetcher` ](https://github.com/onflow/flow-go/blob/master/engine/verification/fetcher/engine.go) engines in verification node, where ingress events to the `requester` are dispatched to `fetcher` through the [`ChunkDataPackHandler`](https://github.com/onflow/flow-go/blob/master/engine/verification/fetcher/requester.go#L20) interface. You may define a similar interface-based interaction between  `splitter.Engine` and `synchronization.RequestHandlerEngine` or go with the `ProcessLocal` option.

Reply:
A `Conduit` is thread-safe as far as I can tell, so it should be fine to share it among multiple engines. As for in coming messages, it already works the way you describe.

Reply:
@smnzhu I revised my comment, please read it again. I was not concerned about the thread safety of conduit, rather how it is created and utilized.

Reply:
I think there is a lot of value in uniform design patterns: engineers familiar with other parts of the code base will recognize such patterns ,which makes the new code easier to understand. Also, constrains and api-level contracts are much easier follow if they are identical all over the code base, thereby reducing probability of bugs and uncovered edge-cases. 
To follow established design patterns is a guideline, and guidelines are not absolute. Therefore, following or deviating from an established design pattern is ultimately the decision of the engineers. But, if you decide to deviate from an established design pattern, please make sure you have a solid reasoning. For example: if following the design pattern would entail that you write 500 lines of extra business-logic code plus tests, that sounds like a valid reason to me. In contrast, "my solution also works" is not a good reason IMHO, because it creates extra cost in the future (for anyone having to understand your custom solution), without any evidence of benefits. 

**I'd like to request that  deviating from the established design pattern is thoroughly document _in code_.** Specifically, please include comments that explain: what the downsides are of following the established pattern in this particular case and the benefits of your alternative approach. Also, please clearly document what engineers have to be cautious about when working with your custom solution. (if there are no extra considerations, please also document this).
If this documentation is more work than following the established pattern, you should most probably follow the existing pattern instead of making up your own. 

In this particular case, it seems to me that @yhassanzadeh13's suggestion wouldn't entail noteworthy overhead (?) Wouldn't it just be that there is one dedicated engine associated with the conduit and all outgoing messages for this conduit would be routed through that engine? 
Also from a conceptual perspective, having one dedicated engine for a conduit seems very intuitive to me. As explained in my [Coding Conventions for Flow](https://github.com/onflow/flow-go/blob/alex/coding-guidelines/docs/Codingconventions.md): 
> On a high level, a Flow node consists of various data-processing components that receive information from node-internal and external sources. We model the data flow inside a Flow node as a graph: a data-processing component forms a vertex; two components exchanging messages is represented as an edge. An Engine is the interface for a data-processing component.

In other words, a conduit wrapped into an engine fits very organically in this data flow graph. But if we start moving  data within a node outside of the engine abstraction, it becomes much harder to comprehend. Sooner or later, we will have to draw the internal data flow graph for each node role. For some nodes, like consensus nodes, this data flow graph is going to be pretty complicated when using engines as vertices alone. Now imagine, we have different types of data flow vertices (engine, conduits, ...), which would add even more complexity to the mental mode. 

I guess the bottom line of my argument is this:
* We are dealing with a system here, that has extremely high intrinsic complexity. Engineers are generally able to understand any small piece of code on isolation. But the challenge for an engineer is generally a wholistic picture of the broader system. 
* Therefore, I think it is most important to keep the conceptual design of the overall system as clean as possible. Specifically, I would encourage you to avoid different types of vertices in the data flow graph, such as engines and conduits.  
* I feel it is a good trade of to add a bit more boilerplate code such that the conceptual design and mathematical abstractions are easily recognizable by engineers. The engineers are not going to be confused by the little extra boilerplate code; but they are going to get completely lost in all the details if they can't identify overarching design patterns. 
* Likewise, a some extra boilerplate code is irrelevant for mathematical correctness or liveness proofs. But a non-uniform conceptual design is going to make such proofs much harder.

Reply:
Okay, I've updated it to use a different approach in e9bafdb and subsequent commits.

Reply:
🙇 🖖